src/sink/amqp.rs

```diff
@@ -0,0 +1,268 @@
+        })
+    }
+}
+
+impl ConfigImpl for Config {}
+
+/// Amqp offramp connector
+pub struct Amqp {
+    sink_url: TremorUrl,
+    config: Config,
+    postprocessors: Postprocessors,
+    reply_channel: Sender<sink::Reply>,
+    channel: Option<Channel>,
+    error_rx: Receiver<()>,
+    error_tx: Sender<()>,
+}
+
+impl fmt::Debug for Amqp {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "[Sink::{}] RoutingKey: {}", &self.sink_url, self.config.routing_key)
+    }
+}
+
+impl offramp::Impl for Amqp {
+    fn from_config(config: &Option<OpConfig>) -> Result<Box<dyn Offramp>> {
+        if let Some(config) = config {
+            let config: Config = Config::new(config)?;
+            let (dummy_tx, _) = bounded(1);
+            let (error_tx, error_rx) = bounded(crate::QSIZE);
+            Ok(SinkManager::new_box(Self {
+                sink_url: TremorUrl::from_offramp_id("amqp")?, // dummy
+                config,
+                postprocessors: vec![],
+                reply_channel: dummy_tx,
+                channel: None,
+                error_rx,
+                error_tx,
+            }))
+        } else {
+            Err("Amqp offramp requires a config".into())
+        }
+    }
+}
+
+impl Amqp {
+    fn handle_channel(&mut self) -> Result<Option<&Channel>> {
+        while let Ok(()) = self.error_rx.try_recv() {
+            self.channel = None;
+        }
+        if self.channel.is_none() {
+            self.channel = Some(self.config.channel()?);
+            return Ok(self.channel.as_ref());
+        }
+        Ok(None)
+    }
+}
+
+#[async_trait::async_trait]
+impl Sink for Amqp {
+    async fn on_event(
+        &mut self,
+        _input: &str,
+        codec: &mut dyn Codec,
+        _codec_map: &HashMap<String, Box<dyn Codec>>,
+        mut event: Event,
+    ) -> ResultVec {
+        self.handle_channel()?;
+        let ingest_ns = event.ingest_ns;
+        let processing_start = Instant::now();
+        /*
+        // evaluate here to avoid borrowing again while borrowed.
+        let config_reply = self.config.reply.as_deref();
+        let op_meta = &event.op_meta;
+        self.merged_meta.merge(op_meta.clone());
+        */
+        let insight_event = event.insight_ack();
+        if let Some(channel) = &mut self.channel {
+            for (value, _) in event.value_meta_iter() {
+                let encoded = codec.encode(value)?;
+                let processed =
+                    postprocess(self.postprocessors.as_mut_slice(), ingest_ns, encoded)?;
+                //let headers = meta.get("nats").and_then(|v| v.get_object("headers"));
+                for payload in processed {
+                    /*
+                    // prepare message reply
+                    let message_reply = reply.or(config_reply);
+                    */
+                    // prepare message headers
+                    let properties = BasicProperties::default();
+                    /*
```

Comment:
Hades would be really nice, but not a blocker, lets throw in a ticket to add them later  if you're OK with that

Reply:
I've no idea what I mean with this so I'll close it :joy:

Reply:
Headers postponed for now due to lack of time, sry.